### DISCUSSION TITLE
Metal geometry limits

### DISCUSSION THREAD

**Discussion by tottero** [google/filament#4113] [Q&A]
Hi,

I can repro the following Metal error consistently if I add a few hundred objects as separate renderables which are highly tessellated:

`Execution of the command buffer was aborted due to an error during execution. Too much geometry to support memoryless render pass attachments. (IOAF code 13)`

The memory usage is not that high, so it is most likely a different constraint. The upper limit of the number of the spheres that can be rendered without the error is always the same on the device. However, if I import a gltf scene instead which contains a lot more from these objects then everything behaves as expected. Would you mind giving some insights on what can trigger these Metal errors? Is there a difference in how the gltfio importer sets up the data or the buffers maybe?

**Comment by bejado** [comment]
Can you link the glTF file that you're using to reproduce?

**Reply by tottero** [reply]
Yeah sure:
[https://drive.google.com/file/d/1ZTkjOSI_Zy775eLO6vZanSGrP_j4HMsg/view?usp=sharing](url)
There are 300 spheres, each ~10k triangles, no errors. If I add even 200 of these 10k spheres as separate renderables the error occurs.

**Reply by tottero** [reply]
Just to clarify, reading gltf files are robust but creating geometry procedurally in the code from index/vertex/tangent arrays might cause these Metal errors when adding them as separate renderables. The gltfio uses buffer objects, can that have a positive impact?

**Reply by bejado** [reply]
Yes, you could try using buffer objects. Can you show how you're creating the spheres as separate renderables?

**Reply by romainguy** [reply]
@tottero `Renderable` objects are fairly cheap (they are basically a buffer of uniforms and a draw call). A good optimization is indeed to batch geometry in fewer `Index`/`VertexBuffer` objects and have multiple `Renderable` instances point at those shared buffers.

**Reply by tottero** [reply]
Thanks! In my case the imported scene is not static, objects can be added or removed. How would you deal with deleting individual objects if the underlying allocated buffer data is shared?

**Reply by romainguy** [reply]
Do you need the vertices to be removed from memory? You can just remove the renderables from the scene.

**Reply by tottero** [reply]
I tried using a single index and vertex buffer with offset for multiple renderables but it did not resolve the memoryless render pass Metal error. 
However, it is really straight-forward to reproduce the problem in the hello-triangle iOS example. You can keep adding the same triangle vertices in a loop to generate more complex geometry (~10-20K triangles). Instead of having a single renderable, create a few hundred renderables as unique buffers:

```
struct App {
    VertexBuffer* vb[COUNT];
    IndexBuffer* ib[COUNT];
    Material* mat;
    Entity renderable[COUNT];
};
```

If you enable AA with setSampleCount() then the error appears consistently and white frames are rendered.
